core/shared/src/main/scala/fs2/Scan.scala

```diff
@@ -0,0 +1,242 @@
+      (s2, acc ++ os)
+    }
+
+  /** Completion function. */
+  def onComplete(s: S): Chunk[O] = onComplete_(s)
+
+  /** Converts this scan to a pipe. */
+  def toPipe[F[_]]: Pipe[F, I, O] =
+    _.pull
+      .scanChunks(initial)(transformAccumulate)
+      .flatMap(state => Pull.output(onComplete(state)))
+      .stream
+
+  /** Steps this scan by a single input, returning a new scan and the output elements computed from the input. */
+  def step(i: I): (Scan[S, I, O], Chunk[O]) = {
+    val (s, os) = transform(initial, i)
+    (new Scan(s, transform_, onComplete_), os)
+  }
+
+  /** Composes the supplied scan with this scan.
+    *
+    * The resulting scan maintains the state of each of the input scans independently.
+    */
+  def andThen[S2, O2](that: Scan[S2, O, O2]): Scan[(S, S2), I, O2] =
+    Scan[(S, S2), I, O2]((initial, that.initial))(
+      { case ((s, s2), i) =>
+        val (sp, os) = transform(s, i)
+        val (s2p, out) = that.transformAccumulate(s2, os)
+        ((sp, s2p), out)
+      },
+      { case (s, s2) =>
+        val (s3, out) = that.transformAccumulate(s2, onComplete(s))
+        out ++ that.onComplete(s3)
+      }
+    )
+
+  /** Returns a new scan which transforms output values using the supplied function. */
+  def map[O2](f: O => O2): Scan[S, I, O2] =
+    new Scan(
+      initial,
+      transform_.andThen[(S, Chunk[O2])] { case (s, os) => (s, os.map(f)) },
+      onComplete_.andThen(_.map(f))
+    )
+
+  /** Returns a new scan which transforms input values using the supplied function. */
+  def contramap[I2](f: I2 => I): Scan[S, I2, O] =
+    new Scan(
+      initial,
+      AndThen[(S, I2), (S, I)] { case (s, i2) => (s, f(i2)) }.andThen(transform_),
+      onComplete_
+    )
+
+  /** Transforms the state type. */
+  def imapState[S2](g: S => S2)(f: S2 => S): Scan[S2, I, O] =
+    Scan[S2, I, O](g(initial))(
+      { (s2, i) =>
+        val (s3, os) = transform(f(s2), i)
+        (g(s3), os)
+      },
+      AndThen(f).andThen(onComplete_)
+    )
+
+  /** Returns a new scan with transformed input and output types.
+    *
+    * Upon receiving an `I2`, `get` is invoked and the result is fed to the
+    * original scan. For each output value, `set` is invoked with the original
+    * `I2` input and the computed `O`, yielding a new output of type `O2`.
+    */
+  def lens[I2, O2](get: I2 => I, set: (I2, O) => O2): Scan[S, I2, O2] =
+    Scan[S, I2, O2](initial)(
+      { (s, i2) =>
+        val (s2, os) = transform(s, get(i2))
+        (s2, os.map(s => set(i2, s)))
+      },
+      _ => Chunk.empty
+    )
+
+  /** Returns a scan that inputs/outputs pairs of elements, with `I` and `O` in the first element of the pair. */
+  def first[A]: Scan[S, (I, A), (O, A)] =
+    lens(_._1, (t, o) => (o, t._2))
+
+  /** Returns a scan that inputs/outputs pairs of elements, with `I` and `O` in the second element of the pair. */
+  def second[A]: Scan[S, (A, I), (A, O)] =
+    lens(_._2, (t, o) => (t._1, o))
+
```

Comment:
The use of `first`, `second`, `left`, and `right`, make me wonder if we could define a `cats.arrow.Arrow` instance for this datatype, and use the nomenclature from those typeclasses.

Reply:
I added a `Strong` instance but `Arrow`, `Compose`, `Category`, and `Choice` instances are not possible due to the initial state that's kept in a `Scan`.